pubsub/producer.go

```diff
@@ -0,0 +1,331 @@
+// Package pubsub implements publisher/subscriber model (one to many).
+// During normal operation, publisher returns "Promise" when publishing a
+// message, which will return resposne from consumer when awaited.
+// If the consumer processing the request becomes inactive, message is
+// re-inserted (if EnableReproduce flag is enabled), and will be picked up by
+// another consumer.
+// We are assuming here that keeepAliveTimeout is set to some sensible value
+// and once consumer becomes inactive, it doesn't activate without restart.
+package pubsub
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/go-redis/redis/v8"
+	"github.com/google/uuid"
+	"github.com/offchainlabs/nitro/util/containers"
+	"github.com/offchainlabs/nitro/util/redisutil"
+	"github.com/offchainlabs/nitro/util/stopwaiter"
+	"github.com/spf13/pflag"
+)
+
+const (
+	messageKey   = "msg"
+	defaultGroup = "default_consumer_group"
+)
+
+// Generic marshaller for Request and Response generic types.
+// Note: unexported fields will be silently ignored.
+type jsonMarshaller[T any] struct{}
+
+// Marshal marshals generic type object with json marshal.
+func (m jsonMarshaller[T]) Marshal(v T) []byte {
+	data, err := json.Marshal(v)
+	if err != nil {
+		log.Error("error marshaling", "value", v, "error", err)
+		return nil
+	}
+	return data
+}
+
+// Unmarshal converts a JSON byte slice back to the generic type object.
+func (j jsonMarshaller[T]) Unmarshal(val []byte) (T, error) {
```

Comment:
you don't need Marshaller at all. You can just call json.Marshal/Unmarshal directly from producer/consumer

Reply:
Done.